module.js

```diff
@@ -82,8 +76,10 @@ instance.prototype.ChangePlayState = function(action,device) {
             }
         }
     }, function(err) {
-        errorCheck(err)
-        self.ChangePlayState(action);
+        if(errorCheck(err)) {
```

Comment:
errorCheck returns either a nothing or false, which will cause undesired behaviour here as that will always fail the if statement.
This is because the retuns on L40 and L44 are inside the those other functions

If you change L50 to be `return Promise.resolve(false)` and add `return` to the start of L37 so that errorCheck always returns a promise. You can then change this line to do a `errorCheck(err).then(function(ok){` to handle the resulting value.
The same will need to be done for all other uses of errorCheck

Reply:
I did the whole thing and then forgot to add the returns in error check 🤦

Reply:
Would doing this do the trick? It's a bit dirtier but should do the trick, I'm used to languages where a `return` is a break and while always fire, even if nested.

```js
instance.prototype.errorCheck = function(err){
    var self = this; 
    var returnVal = false; 
    //Error Code 401 represents out of date token
    if (err.statusCode == '401') {
        self.spotifyApi.refreshAccessToken().then(
            function(data) {
              self.spotifyApi.setAccessToken(data.body['access_token']);
              returnVal = true;
            },
            function(err) {
                console.log('Could not refresh access token', err);
                returnVal = false;
            }
        );
    }
    else {
        console.log('Something went wrong with an API Call: '+ err);
        returnVal = false;
    }
    return returnVal;
}
```

Reply:
no because refreshAccessToken is asynchronous. so the functions passed into `.then(` will be called many milliseconds after errorCheck has returned a value.

refreshAccessToken is returning a [promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise), the functions given to `.then(` will be executed when the promise 'resolves' (determines the value, or throws an error), so you need to continue with that promise flow here

Reply:
Ahh ok, Promises are confusing. I'll implement it properly then tomorrow.